source/extensions/transformers/aws_lambda/api_gateway_transformer.cc

```diff
@@ -0,0 +1,197 @@
+#include "source/extensions/transformers/aws_lambda/api_gateway_transformer.h"
+
+#include "source/common/http/headers.h"
+#include "source/common/http/utility.h"
+#include "source/common/common/base64.h"
+
+#include "source/common/http/header_map_impl.h"
+
+#include "nlohmann/json.hpp"
+using json = nlohmann::json;
+
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace AwsLambda {
+
+HttpFilters::Transformation::ResponseTransformerConstSharedPtr
+ApiGatewayTransformerFactory::createResponseTransformer(
+    const Protobuf::Message &config,
+    Server::Configuration::CommonFactoryContext &context) {
+    MessageUtil::downcastAndValidate<const ApiGatewayTransformerProto &>(
+          config, context.messageValidationContext().staticValidationVisitor());
+  return std::make_shared<ApiGatewayTransformer>();
+}
+
+ApiGatewayTransformer::ApiGatewayTransformer(){}
+const Envoy::Http::LowerCaseString ApiGatewayTransformer::AMAZON_ERRORTYPE_HEADER = Envoy::Http::LowerCaseString("x-amzn-errortype");
+
+void ApiGatewayTransformer::format_error(
+  Http::ResponseHeaderMap &response_headers,
+  Buffer::Instance &body,
+  ApiGatewayError &error,
+  Http::StreamFilterCallbacks &stream_filter_callbacks) const {
+    ENVOY_STREAM_LOG(debug, "Returning error with message: " + std::string(error.message), stream_filter_callbacks);
+
+    // clear existing response headers
+    response_headers.clear();
+
+    response_headers.setStatus(error.status_code);
+    response_headers.setContentType("text/plain");
+
+    response_headers.setReferenceKey(AMAZON_ERRORTYPE_HEADER, error.code);
+    body.drain(body.length());
+    auto error_message = absl::StrCat(error.code, ": ", error.message);
+    body.add(error_message);
+    response_headers.setContentLength(body.length());
+  }
+
+void ApiGatewayTransformer::transform(
+    Http::RequestOrResponseHeaderMap &header_map,
+    Http::RequestHeaderMap *,
+    Buffer::Instance &body,
+    Http::StreamFilterCallbacks &stream_filter_callbacks) const {
+      ENVOY_STREAM_LOG(debug, "Transforming response", stream_filter_callbacks);
+      // we know we are on the repsonse path, so downcast RequestResponseHeaderMap to ResponseHeaderMap
+      Http::ResponseHeaderMap* response_headers = static_cast<Http::ResponseHeaderMap*>(&header_map);
```

Comment:
I think this cast should be done before the interface receives the `header_map` when implementing the `transform()` function